### PR TITLE
enable isolated build (PEP-517 for tox)

### PR DESCRIPTION
**Pull request by gaborbernat**
https://tox.readthedocs.io/en/latest/example/package.html#packaging 👍

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -3,3 +3,5 @@ requires = [
     "setuptools >= 30.0.2",
     "wheel >= 0.29.0"
 ]
+
+build-backend = "setuptools.build_meta"
```
Comment (msullivan) [inline pyproject.toml]: Add a newline?

**Comment by msullivan** [comment]
I don't have a super clear understanding of what this actually changes?

File: pyproject.toml (hunk 1 above)
Comment (gaborbernat) [reply pyproject.toml]: ok

**Comment by gaborbernat** [comment]
@msullivan you would need to read the PEP-517 and PEP-518 to understand clearly, but simply put ensures that the package build dependencies specified in ``pyproject.toml`` are satisfied before packaging the library.

**Comment by gaborbernat** [comment]
Any chance of a merge on this?

**Comment by gaborbernat** [comment]
Without this, the tox packaging build will silently ignore the pyproject.toml and use whatever the tox host environment has, so for users who have an old wheel/setuptools laying around the tox builds might possibly fail.

In parallel, although pip, for now, does default to using setuptools with PEP-517 if the key is not specified we should note that in the PEP-517 this default is not mandated, so it's not guaranteed to work with all build frontends (and pip itself can decide to not activate PEP-517 if this key is missing). This is in the spirit of explicit is better than implicit, in the true spirit of the Python zen. 

Long story short while this might not be a must have, it's definitely a should have.

**Comment by gvanrossum** [comment]
I'm sorry, I still don't follow. If you're asking us to read and understand PEP 517/518, you will have to wait forever, so please bear with us and help us understand, and hopefully we'll feel comfortable that this PR solves a problem for more people than just you yourself.

For context, while I am an avid pip user, I don't understand much about what it does under the covers any more, and I don't understand at all what tox does or why. (At some point someone convinced us to use it in our .travis.yml and appyeyor.yml files, but I don't recall why.)

Given that context, can you explain what a "tox packaging build" is? I see our pyproject.toml but I don't know what it means -- like so much in the packaging world, it looks like pure cargo-cult to me: "don't ask questions, just put this there else 'things' won't work". Also, what is the "key"  you are referring to? And what is a "build frontend"?

Finally, don't go quoting the Zen of Python to me. Following rules without understanding was never meant to be part of it.

**Comment by gaborbernat** [comment]
I'm sorry, my bad, I assumed you would be more familiar with the PEP-517/518. 
I'll write up a ``short``-ish summary of what the problems without PEP-517/518 and get back to you in a few days. Thanks for your patiance and undestanding!

**Comment by gaborbernat** [comment]
Hello, thanks for being patient with me on this, I wrote a blog post explaining what happens more in detail: https://www.bernat.tech/pep-517-518/ Hope this clears up things. Thanks!

**Comment by msullivan** [comment]
Hm, it seems like this will break pip install from the source directory, right?

**Comment by gaborbernat** [comment]
Not if we ammend to prepend the sys path at the start of setup.py 😎

**Comment by msullivan** [comment]
OK, can you update this PR to do that?

**Comment by gaborbernat** [comment]
yeah will do 😊

**Comment by gaborbernat** [comment]
@msullivan addressed, ready for review 👍

**Comment by msullivan** [comment]
@gaborbernat I tried a `pip install .` from the mypy directory and it is still busted.
```
(asdf) msullivan@enki:~/src/mypy [isolated+]
$ pip install .
Processing /home/msullivan/src/mypy
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  Complete output from command /home/msullivan/src/mypy/asdf/bin/python3.7 /home/msullivan/src/mypy/asdf/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py get_requires_for_build_wheel /tmp/tmp439qg31b:
  Traceback (most recent call last):
    File "/home/msullivan/src/mypy/asdf/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py", line 207, in <module>
      main()
    File "/home/msullivan/src/mypy/asdf/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py", line 197, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/home/msullivan/src/mypy/asdf/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py", line 54, in get_requires_for_build_wheel
      return hook(config_settings)
    File "/tmp/pip-build-env-f4k7nrmc/overlay/lib/python3.7/site-packages/setuptools/build_meta.py", line 130, in get_requires_for_build_wheel
      return self._get_build_requires(config_settings, requirements=['wheel'])
    File "/tmp/pip-build-env-f4k7nrmc/overlay/lib/python3.7/site-packages/setuptools/build_meta.py", line 112, in _get_build_requires
      self.run_setup()
    File "/tmp/pip-build-env-f4k7nrmc/overlay/lib/python3.7/site-packages/setuptools/build_meta.py", line 126, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 20, in <module>
      from mypy.version import __version__ as version
  ModuleNotFoundError: No module named 'mypy'
  
  ----------------------------------------
Command "/home/msullivan/src/mypy/asdf/bin/python3.7 /home/msullivan/src/mypy/asdf/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py get_requires_for_build_wheel /tmp/tmp439qg31b" failed with error code 1 in /tmp/pip-req-build-d22a3xxy
```

**Comment by gaborbernat** [comment]
@msullivan indeed, fixed now 👍
